### ISSUE TITLE
NR.5 Example, good | check_size adjustment

### ISSUE THREAD

**Issue by Zyranivia** [isocpp/CppCoreGuidelines#1478]
[Quick Link to NR.5](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rnr-two-phase-init)

In the good example, we have the following contract check:
```
static size_t check_size(size_t s)
{
    // invariant check
    Expects(s > 0);
    return s;
}
```
The problem with this function is, that `check_size(-1)` does not throw as you’d expect only reading the contract.

I propose to implement exactly one of the two ideas below to fix this:
Idea 1: Change the check to `Expects(s != 0);`, as `s` cannot be negative, and thus be more expressive in what exactly is checked.
Idea 2: Change the type of `s` to a signed integer one, maybe `ptrdiff_t` or `intmax_t`, and cast `s` to `size_t` after the check. (Alternatively, change to type of `Picture::mx` and `Picture::my` altogether.)

The latter approach has the advantage that calling `Picture picture(100, -1);` does not try to allocate excessive amounts of memory. On the other hand, the constructor would have to change, too, and a compiler should warn about the signed/unsigned mismatch anyway …

**Comment by scraimer** [comment]
The only correct place to catch this is at the call site, and only with compiler warnings.

Inside `check_size`, all the variables' types are unsigned, so it's too late.

At the call site, the argument is still signed, and it is the conversion to unsigned which is the source of the trouble.

So I'd say this would never be an issue if you compile with the right compiler flags. In GCC and clang that would be `-Wsign-conversion` (see the output at https://gcc.godbolt.org/z/BTckbg)

**Comment by Zyranivia** [comment]
My main concern is the in my eyes misleading contract there. When I see `Expects(s > 0);`, I expect the function to handle negative numbers without relying on an additional compiler warning. I’m new to contracts though, so I’m not sure whether my expectations are what they should be.

**Comment by jwakely** [comment]
I agree that the example could be improved. Using an unsigned type to express the invariant that the arguments are not negative is a bad idea.

**Comment by hsutter** [comment]
Editors' call: We agree the examples should likely use signed integers.
